tests/end_to_end/end_to_end_lib.py

```diff
@@ -0,0 +1,275 @@
+# 1) The number of master servers (default = 1)
+# 2) The number of chunk servers (default = 3)
+# 3) The block size (default = 64MB), and minimum disk space (default = 100 MB)
+# 4) The timeout for grpc (default = 10s), lease (default = 60s), heartbeat (
+#    default = 30s) and client_cache (default = 10m)
+# This function automatically assign ports that are available on the current 
+# machine and automatically generates server names using the "master_server_" +
+# {server_id} and "chunk_server_" + {server_id} as name schemes for the servers
+# Furthermore, all IP addresses from the dns_lookup_table is 0.0.0.0 as most
+# end-to-end tests are expected to run on the same machine
+# 
+# This function also returns a config object in json (as a directory), which
+# will be used later when launching the servers. This is more of an artifact 
+# due to the limitation of importing yaml to Bazel sandbox (running out of 
+# time to figure this out)
+def generate_config_and_config_file(config_filename, num_of_master_server, 
+    num_of_chunk_server, block_size_mb, min_free_disk_space_mb,
+    grpc_timeout_s, lease_timeout_s, heartbeat_timeout_s,
+    client_cache_timeout_m):
+ 
+    config_data = {}
+    config_data["version"] = "1.0"
+    # Config the servers and network sections
+    config_data["servers"] = {}
+    config_data["servers"]["master_servers"] = []
+    config_data["servers"]["chunk_servers"] = []
+    config_data["network"] = {}
+    config_data["network"]["dns_lookup_table"] = {}
+    
+    for i in range(1, num_of_master_server+1):
+        server_name = master_server_name(i)
+        config_data["servers"]["master_servers"].append(server_name)
+        config_data["network"][server_name] = {"hostname" : server_name,
+            "port" : get_avaialble_port()}
+        config_data["network"]["dns_lookup_table"][server_name] = "0.0.0.0"
+
+    for i in range(1, num_of_chunk_server+1):
+        server_name = chunk_server_name(i)
+        config_data["servers"]["chunk_servers"].append(server_name)
+        config_data["network"][server_name] = {"hostname" : server_name, 
+            "port" : get_avaialble_port()}
+        config_data["network"]["dns_lookup_table"][server_name] = "0.0.0.0"
+  
+    # Generate disk config
+    config_data["disk"] = {"block_size_mb" : block_size_mb,
+                           "min_free_disk_space_mb" : 
+                               min_free_disk_space_mb}
+    # Generate timeout config
+    config_data["timeout"] = {"grpc" : str(grpc_timeout_s) + "s",
+        "lease" : str(lease_timeout_s) + "s", 
+        "heartbeat" : str(heartbeat_timeout_s) + "s",
+        "client_cache" : str(client_cache_timeout_m) + "m"}
+ 
+    # TODO(Xi): Ideally we should just dump the dictionary into text using
+    # the yaml module. But I ran into some issues in Bazel setting it up
+    # If time allows (most likely not), will refactor this
+    indent = " "
```

Comment:
nit. you prob can use a `\t` tab character, and the python will write it correctly

Reply:
Sounds good.

Reply:
hmm looks like there are some issues with the tab char somehow (see error message as "invalid map value"). Did not bother looking into this extensively. Let's just leave it as is since it worked.